File: code/python/hash/separate_chaining_hash.py
from sequential_search import SequentialSearch


class SeparateChainingHash:
    INIT_CAPACITY = 4

    def __init__(self, cap=INIT_CAPACITY):
        self._m = cap
        self._n = 0
        self._table = [None] * cap
        for i in range(0, cap):
            self._table[i] = SequentialSearch()

    def size(self):
        return self._n

    def is_empty(self):
        return self._n == 0

    def _resize(self, cap):
        t = SeparateChainingHash(cap)
        for item in self._table:
            for (k, v) in item.items():
                t[k] = v
        self._m = t._m
        self._table = t._table

    def _hash(self, key):
        return hash(key) % self._m

    def __getitem__(self, key):
        if key is None:
            raise ValueError
        i = self._hash(key)
        return self._table[i][key]

    def contains(self, key):
        if key is None:
            raise ValueError
        return self[key] is not None

    def __setitem__(self, key, value):
        if key is None:
            raise ValueError
        if value is None:
            del self[key]
            return
        if self._n >= 10 * self._m:
            self._resize(2 * self._m)
        i = self._hash(key)
        if not self._table[i].contains(key):
            self._n += 1
        self._table[i][key] = value

    def __delitem__(self, key):
        if key is None:
            raise ValueError
        i = self._hash(key)
        if self._table[i].contains(key):
            self._n -= 1
        del self._table[i][key]

        if self._m > self.INIT_CAPACITY and self._n < 2 * self._m:
            self._resize(self._m // 2)

File: code/java/hash/src/main/java/org/swufe/datastructures/SeparateChainingHash.java
package org.swufe.datastructures;

public class SeparateChainingHash<Key, Value> {

    private static final int INIT_CAPACITY = 4;
    private int m;
    private int n;
    private SequentialSearch<Key, Value>[] table;

    @SuppressWarnings("unchecked")
    public SeparateChainingHash(int cap) {
        m = cap;
        n = 0;
        table = (SequentialSearch<Key, Value>[]) new SequentialSearch[cap];
        for (int i = 0; i < m; i++) {
            table[i] = new SequentialSearch<>();
        }
    }

    public SeparateChainingHash() {
        this(INIT_CAPACITY);
    }

    public int size() {
        return n;
    }

    public boolean isEmpty() {
        return n == 0;
    }

    private int hash(Key key) {
        int h = key.hashCode();
        h ^= (h >>> 16);
        return h & (m - 1);
    }

    public void put(Key key, Value value) {
        if (key == null) throw new IllegalArgumentException();
        if (value == null) {
            delete(key);
            return;
        }
        if (n >= 10 * m) resize(2 * m);

        int i = hash(key);
        if (!table[i].contains(key)) n++;
        table[i].put(key, value);
    }

    public Value get(Key key) {
        if (key == null) throw new IllegalArgumentException();
        int i = hash(key);
        return table[i].get(key);
    }

    public boolean contains(Key key) {
        if (key == null) throw new IllegalArgumentException();
        return get(key) != null;
    }

    public void delete(Key key) {
        if (key == null) throw new IllegalArgumentException();

        int i = hash(key);
        if (table[i].contains(key)) n--;
        table[i].delete(key);

        if (m > INIT_CAPACITY && n <= 2 * m) resize(m / 2);
    }

    private void resize(int cap) {
        SeparateChainingHash<Key, Value> t = new SeparateChainingHash<>();
        for (int i = 0; i < m; i++) {
            for (Key key : table[i].keys()) {
                // it will boost the performance
                // if key and its associated value are returned
                t.put(key, table[i].get(key));
            }
        }
        m = t.m;
        table = t.table;
    }

}
